node.js

```diff
@@ -96,98 +99,211 @@ Node.prototype.addChild = function (node) {
   return this
 }
 
-Node.prototype.reset = function (prefix, versions) {
+Node.prototype.reset = function (prefix, constraints) {
   this.prefix = prefix
   this.children = {}
+  this.handlers = []
   this.kind = this.types.STATIC
-  this.handler = null
   this.numberOfChildren = 0
   this.regex = null
   this.wildcardChild = null
-  this.versions = versions
+  this.hasConstraints = true
+  this._decompileGetHandlerMatchingConstraints()
   return this
 }
 
+Node.prototype.split = function (length) {
+  const newChild = new Node(
+    {
+      prefix: this.prefix.slice(length),
+      children: this.children,
+      kind: this.kind,
+      handlers: this.handlers.slice(0),
+      regex: this.regex,
+      constrainer: this.constrainer,
+      hasConstraints: this.hasConstraints
+    }
+  )
+
+  if (this.wildcardChild !== null) {
+    newChild.wildcardChild = this.wildcardChild
+  }
+
+  this.reset(this.prefix.slice(0, length))
+  this.addChild(newChild)
+  return newChild
+}
+
 Node.prototype.findByLabel = function (path) {
   return this.children[path[0]]
 }
 
-Node.prototype.findChild = function (path) {
+Node.prototype.findMatchingChild = function (derivedConstraints, path) {
   var child = this.children[path[0]]
-  if (child !== undefined && (child.numberOfChildren > 0 || child.handler !== null)) {
+  if (child !== undefined && (child.numberOfChildren > 0 || child.getMatchingHandler(derivedConstraints) !== null)) {
```

Comment:
you are calling getMatchingHandler 3 times in this function, this is probably superflous.

Reply:
The `child` changes for each call, it's the piece that implements static > param > wildcard route checking